gpcontrib/gg_tables_tracking/src/track_files.c

```diff
@@ -0,0 +1,1560 @@
+	SetConfigOption("gg_tables_tracking.tracking_relams",
+					buf.data,
+					PGC_SUSET, PGC_S_DATABASE);
+
+	list_free_deep(seen_relams);
+	pfree(buf.data);
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_trigger_initial_snapshot(PG_FUNCTION_ARGS)
+{
+	Oid			dbid = get_dbid(PG_GETARG_OID(0));
+	bloom_op_ctx_t ctx = {0};
+
+	tf_check_shmem_error();
+
+	if (Gp_role != GP_ROLE_DISPATCH && IS_QUERY_DISPATCHER())
+	{
+		ereport(ERROR,
+				(errmsg("Cannot execute tracking_trigger_initial_snapshot outside query dispatcher")));
+	}
+
+	ereport(LOG,
+			(errmsg("[gg_tables_tracking] tracking_trigger_initial_snapshot dbid: %u", dbid)));
+
+	ctx = bloom_set_get_entry(MyDatabaseId, LW_SHARED, LW_EXCLUSIVE);
+
+	if (!ctx.entry)
+	{
+		bloom_set_release(&ctx);
+		ereport(ERROR,
+		(errmsg("Failed to find corresponding filter to database %u", dbid)));
+	}
+
+	if (Gp_role == GP_ROLE_DISPATCH && !pg_atomic_unlocked_test_flag(&ctx.entry->capture_in_progress))
+	{
+		bloom_set_release(&ctx);
+		ereport(ERROR,
+		  (errmsg("Cannot modify track during track acquisition %u", dbid)));
+	}
+
+	bloom_set_all(&ctx.entry->bloom);
+	bloom_set_release(&ctx);
+
+	if (Gp_role == GP_ROLE_DISPATCH)
+	{
+		char	   *cmd = psprintf("select gg_tables_tracking.tracking_trigger_initial_snapshot(%u)", dbid);
+
+		CdbDispatchCommand(cmd, 0, NULL);
+	}
+
+	PG_RETURN_BOOL(true);
+}
+
+Datum
+tracking_is_initial_snapshot_triggered(PG_FUNCTION_ARGS)
+{
+	Oid			dbid = get_dbid(PG_GETARG_OID(0));
+	bool		is_triggered = false;
+
+	tf_check_shmem_error();
+
+	is_triggered = bloom_set_is_all_bits_triggered(dbid);
+
+	ereport(LOG,
+			(errmsg("[gg_tables_tracking] is_initial_snapshot_triggered:%d dbid: %u", is_triggered, dbid)));
+
+	PG_RETURN_BOOL(is_triggered);
+}
+
+Datum
+tracking_is_segment_initialized(PG_FUNCTION_ARGS)
+{
+	ReturnSetInfo *rsi;
+	TupleDesc	tupdesc;
+	HeapTuple	tuple;
+	Datum		values[2];
+	bool		nulls[2] = {false, false};
+	Datum		result;
+
+	tf_check_shmem_error();
+
+	rsi = (ReturnSetInfo *) fcinfo->resultinfo;
+	tupdesc = rsi->expectedDesc;
+
+	/* Populate an output tuple. */
+	values[0] = Int32GetDatum(GpIdentity.segindex);
+	values[1] = BoolGetDatum(pg_atomic_unlocked_test_flag(&tf_shared_state->tracking_is_initialized) == false);
+
+	tuple = heap_form_tuple(tupdesc, values, nulls);
+	result = HeapTupleGetDatum(tuple);
+
+	PG_RETURN_DATUM(result);
+}
+
+static bool
+is_explain_analyze(List *options)
+{
+	ListCell   *lc;
+
+	foreach(lc, options)
+	{
+		DefElem    *opt = (DefElem *) lfirst(lc);
+
+		if (pg_strcasecmp(opt->defname, "analyze") == 0)
+		{
+			return defGetBoolean(opt);
+		}
+	}
+	return false;
+}
+
+static void
+explain_detector_ProcessUtility(PlannedStmt *stmt,
+								const char *queryString,
+								ProcessUtilityContext context,
+								ParamListInfo params,
+								QueryEnvironment *queryEnv,
+								DestReceiver *dest,
+								char *completionTag)
+{
+	Node	   *parsetree = stmt->utilityStmt;
+
+	if (parsetree && IsA(parsetree, ExplainStmt))
```

Comment:
I think that checking is unnecessary, but you can make sure like this
```suggestion
	Assert(parsetree != NULL);

	if (IsA(parsetree, ExplainStmt))
```